File: python/navigation/local_planning/tracking/_path.py
"""Reference-path lookahead geometry for tracking planners.

Path trackers (Pure Pursuit and its descendants) additionally need to
intersect a lookahead circle with the polyline -- kept here as a tracking-only
free function. The point-projection/progress-index primitives this family
shares with the band family live in `local_planning/_geometry.py` and are
re-exported below so existing `from ._path import ...` call sites keep
working unchanged. Mirrors the C++ `local_planning/tracking/path.hpp`.
"""

from __future__ import annotations

import math

from navigation.core.types import Point
from navigation.local_planning._geometry import (
    advance_progress_index,
    closest_point_on_segment,
    sq_dist,
)

__all__ = [
    "advance_progress_index",
    "closest_point_on_segment",
    "lookahead_point",
    "segment_circle_forward_t",
    "sq_dist",
]


def segment_circle_forward_t(p: Point, a: Point, b: Point, radius: float) -> float | None:
    """Forward-most intersection of the probe-centered lookahead circle with
    segment a->b, as a parameter t in [0, 1], or None if the segment stays
    entirely inside/outside the circle.

    Solves |a + t*(b-a) - p|^2 = radius^2 (Coulter 1992 sec. 3: circle-line
    intersection) and keeps the larger root in range -- the exit point, i.e.
    further along the path -- so the chosen point always leads the robot
    forward rather than back toward where it entered the circle.
    """
    dx, dy = b[0] - a[0], b[1] - a[1]
    fx, fy = a[0] - p[0], a[1] - p[1]
    aa = dx * dx + dy * dy
    if aa < 1e-12:
        return None
    bb = 2.0 * (fx * dx + fy * dy)
    cc = fx * fx + fy * fy - radius * radius
    disc = bb * bb - 4.0 * aa * cc
    if disc < 0.0:
        return None
    sq = math.sqrt(disc)
    for t in ((-bb + sq) / (2.0 * aa), (-bb - sq) / (2.0 * aa)):
        if 0.0 <= t <= 1.0:
            return t
    return None


def lookahead_point(
    path: tuple[Point, ...],
    start_index: int,
    robot_xy: Point,
    lookahead_distance: float,
) -> Point:
    """First forward intersection of the lookahead circle with the path at or
    after ``start_index``; falls back to the path end when the remaining path
    is shorter than the lookahead distance."""
    for i in range(start_index, len(path) - 1):
        t = segment_circle_forward_t(robot_xy, path[i], path[i + 1], lookahead_distance)
        if t is not None:
            a, b = path[i], path[i + 1]
            return (a[0] + t * (b[0] - a[0]), a[1] + t * (b[1] - a[1]))
    # No segment crosses the lookahead circle -- the remaining path is
    # shorter than L_d, so aim at the path's end (the goal).
    return path[-1]

File: cpp/src/local_planning/tracking/path.cpp
#include "navigation/local_planning/tracking/path.hpp"

#include <algorithm>
#include <cmath>
#include <limits>

namespace navigation::local_planning {

core::Point closest_point_on_segment(const core::Point& p, const core::Point& a,
                                     const core::Point& b) {
  double dx = b.x - a.x, dy = b.y - a.y;
  double seg_len_sq = dx * dx + dy * dy;
  if (seg_len_sq < 1e-12) return a;
  double t = std::max(0.0, std::min(1.0, ((p.x - a.x) * dx + (p.y - a.y) * dy) / seg_len_sq));
  return core::Point{a.x + t * dx, a.y + t * dy};
}

double sq_dist(const core::Point& a, const core::Point& b) {
  double dx = a.x - b.x, dy = a.y - b.y;
  return dx * dx + dy * dy;
}

// Solves |a + t*(b-a) - p|^2 = radius^2 (Coulter 1992 sec. 3: circle-line
// intersection) and keeps the larger root in range -- the exit point, i.e.
// further along the path -- so the chosen point always leads the robot forward
// rather than back toward where it entered the circle.
std::optional<double> segment_circle_forward_t(const core::Point& p, const core::Point& a,
                                               const core::Point& b, double radius) {
  double dx = b.x - a.x, dy = b.y - a.y;
  double fx = a.x - p.x, fy = a.y - p.y;
  double aa = dx * dx + dy * dy;
  if (aa < 1e-12) return std::nullopt;
  double bb = 2.0 * (fx * dx + fy * dy);
  double cc = fx * fx + fy * fy - radius * radius;
  double disc = bb * bb - 4.0 * aa * cc;
  if (disc < 0.0) return std::nullopt;
  double sq = std::sqrt(disc);
  for (double t : {(-bb + sq) / (2.0 * aa), (-bb - sq) / (2.0 * aa)}) {
    if (t >= 0.0 && t <= 1.0) return t;
  }
  return std::nullopt;
}

int advance_progress_index(const std::vector<core::Point>& path, const core::Point& probe,
                           int start_index) {
  if (path.size() < 2) return start_index;
  int best_index = start_index;
  double best_sq_dist = std::numeric_limits<double>::infinity();
  for (int i = start_index; i < static_cast<int>(path.size()) - 1; ++i) {
    core::Point closest = closest_point_on_segment(probe, path[static_cast<size_t>(i)],
                                                   path[static_cast<size_t>(i) + 1]);
    double d = sq_dist(probe, closest);
    // <=, not <: consecutive segments share their joint endpoint, so a probe
    // sitting exactly at a corner ties every segment ending/starting there.
    // Preferring the later (more forward) segment on a tie keeps progress
    // advancing through the corner instead of latching onto the segment just
    // traveled.
    if (d <= best_sq_dist) {
      best_sq_dist = d;
      best_index = i;
    }
  }
  return best_index;
}

core::Point lookahead_point(const std::vector<core::Point>& path, int start_index,
                            const core::Point& robot_xy, double lookahead_distance) {
  // No segment crossing the lookahead circle means the remaining path is
  // shorter than L_d, so aim at the path's end (the goal).
  core::Point target = path.empty() ? robot_xy : path.back();
  for (int i = start_index; !path.empty() && i < static_cast<int>(path.size()) - 1; ++i) {
    auto t = segment_circle_forward_t(robot_xy, path[static_cast<size_t>(i)],
                                      path[static_cast<size_t>(i) + 1], lookahead_distance);
    if (t) {
      const core::Point& a = path[static_cast<size_t>(i)];
      const core::Point& b = path[static_cast<size_t>(i) + 1];
      target = core::Point{a.x + *t * (b.x - a.x), a.y + *t * (b.y - a.y)};
      break;
    }
  }
  return target;
}

}  // namespace navigation::local_planning
